### PR TITLE
Prevent resolutions after calling Dispose

### PR DESCRIPTION
**Pull request by pakrym**

### CODE REVIEW

File: test/DI.Tests/ServiceProviderContainerTests.cs [hunk 1]
```diff
@@ -116,13 +116,65 @@ public void DoesNotDisposeSingletonInstances()
...
+        {
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddScoped<IFakeService, FakeService>();
+
+            var provider = CreateServiceProvider(serviceCollection);
+            var scope = provider.CreateScope();
+            scope.Dispose();
+
+            Assert.Throws<ObjectDisposedException>(() => scope.ServiceProvider.GetService<IFakeService>());
+            //Chec that resolution from root works
```
Comment (davidfowl) [inline test/DI.Tests/ServiceProviderContainerTests.cs]: Check

File: src/DI/ServiceLookup/ServiceProviderEngine.cs [hunk 2]
```diff
@@ -57,16 +86,11 @@ protected ServiceProviderEngine(IEnumerable<ServiceDescriptor> serviceDescriptor
             return _ => null;
         }
 
-        internal object GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (halter73) [inline src/DI/ServiceLookup/ServiceProviderEngine.cs]: Did you compare perf inlined vs non-inlined?

File: src/DI/ServiceLookup/ServiceProviderEngineScope.cs [hunk 3]
```diff
@@ -26,6 +26,11 @@ public ServiceProviderEngineScope(ServiceProviderEngine engine)
 
         public object GetService(Type serviceType)
         {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(IServiceScope));
```
Comment (halter73) [inline src/DI/ServiceLookup/ServiceProviderEngineScope.cs]: Why change patterns from ServiceProviderEngine and throw ODE directly from the method?

File: test/DI.Tests/ServiceProviderContainerTests.cs [hunk 4]
```diff
@@ -116,13 +116,65 @@ public void DoesNotDisposeSingletonInstances()
...
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddScoped<IFakeService, FakeService>();
+
+            var provider = CreateServiceProvider(serviceCollection);
+            var scope = provider.CreateScope();
+            scope.Dispose();
+
+            Assert.Throws<ObjectDisposedException>(() => scope.ServiceProvider.GetService<IFakeService>());
+            //Check that resolution from root works
+            Assert.NotNull(provider.CreateScope());
```
Comment (halter73) [inline test/DI.Tests/ServiceProviderContainerTests.cs]: Nit: Comment doesn't seem to match assertion since you aren't resolving anything from the root.

**Review by halter73 — APPROVED** [review]

File: test/DI.Tests/ServiceProviderContainerTests.cs (hunk 4 above)
Comment (pakrym) [reply test/DI.Tests/ServiceProviderContainerTests.cs]: Create scope resolves scope factory

File: src/DI/ServiceLookup/ServiceProviderEngineScope.cs (hunk 3 above)
Comment (pakrym) [reply src/DI/ServiceLookup/ServiceProviderEngineScope.cs]: Good point.

File: src/DI/ServiceLookup/ServiceProviderEngine.cs (hunk 2 above)
Comment (pakrym) [reply src/DI/ServiceLookup/ServiceProviderEngine.cs]: No, but we usually extract throwing into non inlined methods so their callers are not increasing in size and inline easier.

**Review by davidfowl — APPROVED** [review]

File: test/DI.Tests/ServiceProviderContainerTests.cs (hunk 4 above)
Comment (halter73) [reply test/DI.Tests/ServiceProviderContainerTests.cs]: 👍

**Pull request merged**